File: C++/아기상어.cpp
#include <iostream>
#include <queue>
#include <string.h>
#include <vector>

using namespace std;

typedef struct SHARK {
	int x, y, size, cnt;
}SHARK;

typedef struct FISH {
	int x, y, size, dist;
}FISH;

int dx[4] = { -1,1,0,0 };
int dy[4] = { 0,0,-1,1 };
int input[21][21];
SHARK shark;
int N;

bool cmp(FISH a, FISH b)
{
	if (a.dist < b.dist)
		return true;
	if (a.dist > b.dist)
		return false;
	if (a.y < b.y)
		return true;
	if (a.y > b.y)
		return false;

	if (a.x < b.x)
		return true;
	else
		return false;
}

int solution()
{
	vector<FISH> fishes;
	SHARK s;
	int isVisited[21][21];
	int nx, ny;
	int time = 0;

	while (1)
	{
		memset(isVisited, false, sizeof(isVisited));

		queue<SHARK> q;
		q.push({shark.x, shark.y, 0, 0}); // size 부분 안씀
		isVisited[shark.y][shark.x] = true;
		while (!q.empty())
		{
			s = q.front();
			q.pop();
			for (int i = 0; i < 4; i++)
			{
				nx = s.x + dx[i];
				ny = s.y + dy[i];

				if (nx < 0 || ny < 0 || nx >= N || ny >= N || isVisited[ny][nx])
					continue;
				if (input[ny][nx] == 0 || shark.size == input[ny][nx])
					q.push({ nx, ny, 0, s.cnt + 1 });
				else
				{
					if (shark.size > input[ny][nx])
						fishes.push_back({ nx, ny, input[ny][nx], s.cnt + 1 });
				}
				isVisited[ny][nx] = true;
			}
		}
		if (fishes.size() == 0)
			return time;
		sort(fishes.begin(), fishes.end(), cmp);

		// 우선 순위 높은것 먹음
		input[fishes[0].y][fishes[0].x] = 0;
		time += fishes[0].dist;
		shark.cnt++;
		shark.y = fishes[0].y;
		shark.x = fishes[0].x;
		if (shark.cnt == shark.size)
		{
			shark.cnt = 0;
			shark.size++;
		}
		fishes.clear();
	}
}

int main()
{
	cin >> N;
	for (int i = 0; i < N; i++)
	{
		for (int j = 0; j < N; j++)
		{
			cin >> input[i][j];
			if (input[i][j] == 9)
			{
				shark = { j, i, 2, 0 };
				input[i][j] = 0;
			}
		}
	}
	cout << solution() << endl;
}

File: JAVA/아기상어.java
package practice;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.PriorityQueue;
import java.util.Queue;
import java.util.StringTokenizer;

public class 아기상어 {
	static int N;
	static Fish shark;
	static int time;
	static int[][] space;
	static boolean[][] isVisited = new boolean[N][N];
	static int[] dx = { -1, 0, 1, 0 };
	static int[] dy = { 0, -1, 0, 1 };

	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		N = Integer.parseInt(in.readLine());
		space = new int[N][N];
		time = 0;

		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(in.readLine(), " ");
			for (int j = 0; j < N; j++) {
				space[i][j] = Integer.parseInt(st.nextToken());
				if (space[i][j] == 9) {
					shark = new Fish(2, i, j, 0);
					space[i][j] = 0;
				}
			}
		}

		solution();
		System.out.println(time);
	}

	private static void solution() {
		PriorityQueue<Fish> fishes = new PriorityQueue<>(new Comparator<Fish>() {
			@Override
			public int compare(Fish o1, Fish o2) {
				if (o1.dist > o2.dist) // 거리
					return 1;
				else if(o1.dist < o2.dist)
					return -1;
				else {
					if (o1.y > o2.y)  // 위쪽
						return 1;
					else if(o1.y < o2.y)
						return -1;
					else {
						if (o1.x > o2.x) // 왼쪽
							return 1;
						else
							return -1;
					}
				}
			}
		});
		Queue<Fish> q = new LinkedList<>();
		Fish d_fish;
		Fish tmp;
		int cnt = 0;
		int ny, nx, y, x;

		while (true) {
			isVisited = new boolean[N][N];
			q.add(shark);
			isVisited[shark.y][shark.x] = true;

			while (!q.isEmpty()) {
				tmp = q.poll();
				y = tmp.y;
				x = tmp.x;
				if (space[y][x] != 0 && space[y][x] < shark.size) {
					fishes.offer(tmp);
				}
				for (int i = 0; i < 4; i++) {
					ny = y + dy[i];
					nx = x + dx[i];
					if (ny < 0 || nx < 0 || ny >= N || nx >= N || isVisited[ny][nx])
						continue;
					isVisited[ny][nx] = true;
					if (space[ny][nx] <= shark.size)
						q.offer(new Fish(space[ny][nx], ny, nx, tmp.dist + 1));
				}
			}

			if (fishes.size() == 0)
				return;

			d_fish = fishes.poll();
			time += d_fish.dist;

			shark.y = d_fish.y;
			shark.x = d_fish.x;
			space[d_fish.y][d_fish.x] = 0;
			fishes.clear();
			q.clear();
			cnt++;

			if (shark.size == cnt) {
				shark.size++;
				cnt = 0;
			}
		}
	}

}

class Fish {
	int size;
	int y;
	int x;
	int dist;

	Fish(int size, int y, int x, int dist) {
		this.size = size;
		this.y = y;
		this.x = x;
		this.dist = dist;
	}
}